### PR TITLE
Various smaller improvements related to the development of the Inchworm CTQMC

### PR DESCRIPTION
**Pull request by Wentzell**
All tests passing. Commit message gives good description of each change.
Review each commit separately.

### CODE REVIEW

File: c++/nda/matrix_functions.hpp [hunk 1]
```diff
@@ -24,7 +24,8 @@ namespace nda {
   }
 
   template <CONCEPT(ArrayOfRank<2>) M>
-  get_value_t<M> trace(M const &m) REQUIRES17(is_ndarray_v<M> and (get_rank<M> == 2)) {
+  auto trace(M const &m) REQUIRES17(is_ndarray_v<M>) {
+    static_assert(get_rank<M> == 2);
```
Comment (parcollet) [inline c++/nda/matrix_functions.hpp:28]: Is redondant in C++20,  CONCEPT protects it.
But that is ok.

File: c++/nda/linalg/det_and_inverse.hpp [hunk 2]
```diff
@@ -30,39 +30,40 @@ namespace nda {
...
+  auto determinant_in_place(M &m) REQUIRES(is_matrix_or_view_v<M>) {
+    using value_t = get_value_t<M>;
+    static_assert(std::is_convertible_v<value_t, double> or std::is_convertible_v<value_t, std::complex<double>>,
+	"determinant requires a matrix of values that can be implicitly converted to double or std::complex<double>");
+    static_assert(not std::is_const_v<M>, "determinant_in_place can not be const. It destroys its argument");
+
+    if(m.empty()) return value_t{1};
+
+    if(m.extent(0) != m.extent(1))
+      NDA_RUNTIME_ERROR << "Error in determinant. Matrix is not square but has shape " << m.shape();
```
Comment (parcollet) [inline c++/nda/linalg/det_and_inverse.hpp:43]: consider ASSERT/ ABORT here. It is a logic error, not an exception we should recover from,.

File: c++/nda/mpi/reduce.hpp [hunk 3]
```diff
@@ -37,34 +39,41 @@ namespace nda::lazy_mpi {
...
-        if (std::abs(target_view.data_start() - source.data_start()) < source.size())
-          NDA_RUNTIME_ERROR << "mpi reduce of array : overlapping arrays !";
-      }
+      if constexpr(not mpi::has_mpi_type<value_type>){
 
-      void *v_p       = (void *)target_view.data_start();
-      void *rhs_p     = (void *)source.data_start();
-      auto rhs_n_elem = source.size();
-      auto D          = mpi::mpi_type<value_type>::get();
+	nda::map([this](value_type const & x){ return mpi::reduce(x, this->c, this->root, this->all, this->op); })(source);
```
Comment (parcollet) [inline c++/nda/mpi/reduce.hpp]: correct but `auto const &` is safer coding (you don't need to check that the type is exactly value_type ...)

File: test/c++/nda_mpi.cpp [hunk 4]
```diff
@@ -110,6 +113,29 @@ TEST(Arrays, MPI) { //NOLINT
...
+    a(i) = matrix_t{4, 4};
+    a(i)(k_, l_) << world.rank() * (k_ + l_);
+
+    r_exact(i) = matrix_t{4, 4};
+    r_exact(i)(k_, l_) << (world.size() - 1) * world.size() / 2 * (k_ + l_);
+  }
+
+  arr_t r = mpi::all_reduce(a, world);
+
+  //EXPECT_ARRAY_EQ(r, r_exact); // FIXME Does not compile
```
Comment (parcollet) [inline test/c++/nda_mpi.cpp]: The TEST does not test anything !

File: c++/nda/algorithms.hpp [hunk 5]
```diff
@@ -105,16 +105,13 @@ namespace nda {
...
    */
-  template <typename T>
-  // FIxme IsDoubleOrComplex
-  //require( is_real_or_complex<T>)
-  double frobenius_norm(matrix<T> const &a) {
+  template <CONCEPT(ArrayOfRank<2>) A>
+  double frobenius_norm(A const &a) REQUIRES17(is_ndarray_v<A> &&has_rank<A, 2>) {
     return std::sqrt(fold(
-       [](double r, T const &x) -> double {
+       [](double r, get_value_t<A> const &x) -> double {
```
Comment (parcollet) [inline c++/nda/algorithms.hpp]: Use
auto const &
it is much more readable, one does not need to check that the type is correct, 
i.e. that **no silent conversion will ever occur...**

File: test/c++/nda_algorithms.cpp [hunk 6]
```diff
@@ -73,4 +73,17 @@ TEST(NDA, Algo1) { //NOLINT
...
+TEST(NDA, AlgoMat) { //NOLINT
+  nda::matrix<double> A(3, 3);
+
+  for (int i = 0; i < 3; ++i)
+    for (int j = 0; j < 3; ++j) A(i, j) = std::sqrt(i + 3 * j);
+
+  nda::matrix<double, C_layout, nda::sso<100>> A_SSO(3, 3);
+  A_SSO = A;
+
+  EXPECT_EQ(frobenius_norm(A), std::sqrt(9 * 8 / 2));
```
Comment (parcollet) [inline test/c++/nda_algorithms.cpp:85]: EXPEXT_EQ for double ? 
Shouldn't we put some machine precision ?

File: c++/nda/lapack.hpp [hunk 7]
```diff
@@ -54,8 +54,8 @@ namespace nda::lapack {
    * @param ipiv  Gauss Pivot, cf lapack doc
    *
    */
-  template <typename M>
-  [[nodiscard]] int getrf(M &&m, array<int, 1> &ipiv) {
+  template <typename M, typename CP>
+  [[nodiscard]] int getrf(M &&m, basic_array<int, 1, C_layout, 'A', CP> &ipiv) REQUIRES(is_ndarray_v<std::decay_t<M>> and (get_rank<std::decay_t<M>> == 2)) {
```
Comment (parcollet) [inline c++/nda/lapack.hpp]: We can static_assert instead of REQUIRES here, there is no other option, right ?

File: c++/nda/basic_array.hpp [hunk 8]
```diff
@@ -255,13 +255,13 @@ namespace nda {
 
     //------------------ Factory -------------------------
```
Comment (parcollet) [inline c++/nda/basic_array.hpp:252]: Why this ? 
It should be deduced from the variable passed.

**Review by parcollet — COMMENTED** [review]
All minors point, except zeros<> : I did not understand the point

File: test/c++/nda_mpi.cpp (hunk 4 above)
Comment (Wentzell) [reply test/c++/nda_mpi.cpp]: I actually adjusted the test in a later commit. I will merge these changes together.

File: c++/nda/mpi/reduce.hpp (hunk 3 above)
Comment (Wentzell) [reply c++/nda/mpi/reduce.hpp]: I find that a bit easier to read in fact. It needs to be `value_type`, which is a short name.

File: test/c++/nda_algorithms.cpp (hunk 6 above)
Comment (Wentzell) [reply test/c++/nda_algorithms.cpp:85]: But I expect this to be the exact double to be returned here. It passes.

File: c++/nda/algorithms.hpp (hunk 5 above)
Comment (Wentzell) [reply c++/nda/algorithms.hpp]: Yes I agree here

File: c++/nda/linalg/det_and_inverse.hpp (hunk 2 above)
Comment (Wentzell) [reply c++/nda/linalg/det_and_inverse.hpp:43]: The reason I used an exception here is again the fact that this code may be used through a Python Module.
My understanding is that `std::abort` here would kill also the Python Interpreter!?

File: c++/nda/basic_array.hpp (hunk 8 above)
Comment (Wentzell) [reply c++/nda/basic_array.hpp:252]: The deduction will fail for cases where you provide an initializer list
```c++
 matrix_t::zeros({n, n});
```
That is why I added the default here.

File: c++/nda/linalg/det_and_inverse.hpp (hunk 2 above)
Comment (parcollet) [reply c++/nda/linalg/det_and_inverse.hpp:43]: yes it would.
Ok, fine then.

File: c++/nda/mpi/reduce.hpp (hunk 3 above)
Comment (parcollet) [reply c++/nda/mpi/reduce.hpp]: the danger is that the day that you change the type, and it is not value_type any more 
for any reason, you will impose a **silent copy** 
I think auto const & is more "defensive programming", no bug can happen and the lambda is short.

File: test/c++/nda_mpi.cpp (hunk 4 above)
Comment (parcollet) [reply test/c++/nda_mpi.cpp]: ok, do I need to see this commit ?

File: c++/nda/algorithms.hpp (hunk 5 above)
Comment (parcollet) [reply c++/nda/algorithms.hpp]: why is it different from the previous case ? ok the trait is more complex to read though...
It is a matter of style anyway

File: test/c++/nda_algorithms.cpp (hunk 6 above)
Comment (parcollet) [reply test/c++/nda_algorithms.cpp:85]: as long as it passes...

File: c++/nda/basic_array.hpp (hunk 8 above)
Comment (parcollet) [reply c++/nda/basic_array.hpp:252]: Oh, I see... nice indeed, it make the std::array on the fly.
ok.

File: c++/nda/lapack.hpp (hunk 7 above)
Comment (Wentzell) [reply c++/nda/lapack.hpp]: We could also generalize that to array of rank 2?
static_assert for the rank check is fine with me. In that case I prefer the explicit compiler error in fact.

Comment (parcollet) [reply c++/nda/lapack.hpp]: static_assert for both . why not, there is no other overload ...
Also, only the error in C++20 really matters... C++17 is backward mode.

File: test/c++/nda_mpi.cpp (hunk 4 above)
Comment (Wentzell) [reply test/c++/nda_mpi.cpp]: https://github.com/TRIQS/nda/pull/10/commits/02b5f1c809779190151e1b8e057abb5c27dbdb5b

File: c++/nda/lapack.hpp (hunk 7 above)
Comment (Wentzell) [reply c++/nda/lapack.hpp]: Ok

**Pull request merged**